### ISSUE TITLE
Add a NumPy-like `pad` function

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#60294]
## 🚀 Feature
Add a `torch.pad` function, based on [`numpy.pad`](https://numpy.org/doc/stable/reference/generated/numpy.pad.html)

## Motivation

NumPy compatability.

Plus, this will offer a solution to several issues with `torch.nn.functional.pad`:

* #52205 - reflection padding is only supported if padding width is less than the input's width
* #57911 - circular padding can only wrap once
* #46240 - we don't support symmetric reflection padding
* #28214 and #29863 - we only support a subset of input dimensionality

## Pitch

Most of the `numpy.pad` interface can probably be copied. However, we could potentially extend it to support anything extra that `torch.nn.functional.pad` offers that `numpy.pad` does not. We could also maybe make small improvements upon `numpy.pad`'s interface.

## Alternatives

There are a couple potentially confusing aspects of the `numpy.pad` interface that we could think about improving:
* The `constant_values` and `end_values` arguments can never be used at the same time (they are for different modes), and they logically specify the same exact thing--they're just used slightly differently. We could combine them into one argument.
* The term "end value" (for mode='linear_ramp') is potentially confusing. The word end, in my mind, implies one of two edges of something. For instance, there is one start and one end of a footrace. But in `numpy.pad`, "end" is meant as the outermost edges of the padding. In the footrace example, both points where the race starts and stops would be "end"s, using this definition of the word. "edge value" could be better, but one of the modes is called "edge", so that could cause confusion, too. Perhaps "outer value" would be a good choice.

`numpy.pad` allows you to specify a function for the `mode` argument. I don't know of any PyTorch operations where a function can be given as an argument, so I'm guessing that can't support this feature yet.

**Comment by kurtamohler** [comment]
BTW, I'm having trouble figuring out how `numpy.pad(..., mode='reflect', reflect_type='odd')` works when the padding width is equal to or greater than the input width. It doesn't seem like the documentation fully explains it. Here's an example case:

```python
np.random.seed(0)
a = np.random.randint(0, 10, (5,))
a_padded = np.pad(a, (0, 10), mode='reflect', reflect_type='odd')
print(f'a:\n{a}')
print(f'a_padded:\n{a_padded}')
```

Output:
```python
a:
[5 0 3 3 7]
a_padded:
[ 5  0  3  3  7 11 11 14  9  4  7  7 11 15 15]
```

I tried to figure it out in my notes under the `mode='reflect', reflect_type='odd'` section of my notes [here](https://github.com/kurtamohler/notes/blob/main/numpy-pad-examples/numpy-pad-examples.ipynb), but I didn't find an answer that makes sense.

If someone understands how the padding is being calculated in this case, let me know. Otherwise, I'll probably just look into the code

**Comment by mruberry** [comment]
fyi @jbschlosser to consider if their are `nn.functional.pad` concerns

A NumPy-like pad function that generalizes PyTorch's existing pad functionality is exciting! Maybe we should focus on a few pad modes to start?

Also, I have a factual question. Where PyTorch and NumPy define a padding operation with the same name, like "reflect" or "constant," do they define the same operation today? If not, what's different, and how should we think about resolving the discrepancy? 

Other padding modes in PyTorch are circular (which is divergent between the Python and C++ implementations, if I recall correctly, and would be nice to make consistent), and replication. Some of these have exposures outside of nn.functional.pad, like the ReplicationPad1d module. In the future we might want to think about a more general padding module, too, that uses torch.pad.

**Comment by kurtamohler** [comment]
> Maybe we should focus on a few pad modes to start?

Yeah, that seems like a good idea. 'constant' should be included in the first PR since it's the default. We should probably have at least two modes in the first PR--'empty' seems like a good second choice to me.

> Where PyTorch and NumPy define a padding operation with the same name, like "reflect" or "constant," do they define the same operation today? If not, what's different, and how should we think about resolving the discrepancy?

`numpy.pad` supports all four of the modes that `torch.nn.functional.pad offers (at least in the Python API--I haven't investigated the C++ API yet). The behavior of the corresponding modes seem to match. However the names differ in some cases and the PyTorch function has some restrictions that NumPy does not have, like input dimensionality and pad width. Here's a table of how the modes map to each other and the restrictions I've seen with the PyTorch function:

| numpy.pad mode name | torch.nn.functional.pad mode name | PyTorch restrictions |
| --- | --- | --- |
| 'constant' | 'constant' | Uses only one constant for all padding values. NumPy allows you to specify different constants for each dimension for before- and after-padding values. |
| 'reflect' (with reflect_type='even', default) | 'reflect' | Input has to have 3 or 4 dimensions. Cannot pad the first two dimensions. Padding widths must be less than the size of the corresponding dimension in the input. |
| 'edge' | 'replicate' | Input has to have 3, 4, or 5 dimensions. Cannot pad the first two dimensions. |
| 'wrap' | 'circular' | Input has to have 3, 4, or 5 dimensions. Cannot pad the first two dimensions. Padding widths must be less than or equal to the size of the corresponding dimension of the input |

So I believe that `numpy.pad` covers everything that `torch.nn.functional.pad` provides. However, it is always possible that there might be some weird edge cases where the two give different results. I can write a comprehensive test to find out.

We'll have to decide what to do about the differing mode names though. I think `torch.pad` should use the same mode names that `numpy.pad` uses. We could consider providing the 'replicate' and 'circular' mode names as well, but just alias them to 'edge' and 'wrap'.

**Comment by mruberry** [comment]
Thanks for the awesome analysis, @kurtamohler. I agree that torch.pad should prefer the NumPy names, but if we can sneak the redundant PyTorch names in, too, that'd be great. That would make updating to torch.pad easier for current PyTorch pad users. Also, seems like NumPy's pad is a superset of the functionality that nn.functional.pad offers. That'd be awesome if we can validate it. I guess we'll need some very thorough tests.

**Comment by kurtamohler** [comment]
I wrote a script to test a bunch of equivalent calls between `numpy.pad` and `torch.nn.functional.pad`. There are a lot of cases that `torch.nn.functional.pad` does not support, but so far, all the supported cases in my test give the same result as `numpy.pad`.

Here's the script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/pad/test_numpy_pad_compat.py

**Comment by kurtamohler** [comment]
I'm thinking about what types we should support for the `pad_width`, `stat_length`, `constant_values`, and `end_values` args.

In numpy, these args are all share the same format. They can take only these forms:
* Single number, `constant`
* Single number in a sequence, `(constant,)`
* Pair sequence, `(before, after)`
* Sequence containing one pair sequence, `((before, after),)`
* Sequence containing a pair sequence for each dimension of the input, `((before0, after0), (before1, after1), ...)`

So if the arg is a 2-D sequence, the length of each inner sequence always has to be 2. The sequences can be lists, tuples, or numpy arrays.

As far as I can tell at the moment, it would only make sense for these arguments to be Tensors in `torch.pad`. It would be simple enough to convert a `numpy.pad`-compatible arg to a tensor using `torch.tensor()`. (We could potentially even do the Tensor conversion automatically, though I'm not sure how much effort that would take--I don't think native_functions.yaml has anything that does that at the moment, so maybe it would require some specialized codegen stuff.)

The only possible alternative to a Tensor, as far know, is to have overloads with both `int`s and `int[]`s. But we'd have to have 2^4 = 16 different overloads (2 overloads for each of the 4 arguments) of `torch.pad` to implement it that way. Doesn't seem like a very good idea to me :D

**Comment by mruberry** [comment]
> As far as I can tell at the moment, it would only make sense for these arguments to be Tensors in `torch.pad`. 

What makes you say that?

> It would be simple enough to convert a `numpy.pad`-compatible arg to a tensor using `torch.tensor()`. (We could potentially even do the Tensor conversion automatically, though I'm not sure how much effort that would take--I don't think native_functions.yaml has anything that does that at the moment, so maybe it would require some specialized codegen stuff.)

One trick that @heitorschueroff has used on einsum is to accept the arguments in Python but then translate them to a different internal format or call. This is supported by the jit but means the UX of the C++ API isn't as easy to use as the Python API. I think that's an OK place to start versus offering a worse Python UX, too.

> The only possible alternative to a Tensor, as far know, is to have overloads with both `int`s and `int[]`s. But we'd have to have 2^4 = 16 different overloads (2 overloads for each of the 4 arguments) of `torch.pad` to implement it that way. Doesn't seem like a very good idea to me :D

Maybe we can use the above trick to implement the function in Python and more cleverly call internal functions, though? Take a look at the einsum implementation here:

https://github.com/pytorch/pytorch/blob/4e347f12422aad80e03ae722429a828ee864b13a/torch/functional.py#L158

**Comment by kurtamohler** [comment]
@mruberry, you're right. I agree with prioritizing UX for the Python interface. So we should support the same types that `numpy.pad` supports for the `pad_width`, `constant_values`, etc arguments in Python. `numpy.pad` also accepts numpy arrays for these arguments, not just the native Python sequence types. So for consistency, I think `torch.pad`'s Python interface should also support Tensors for these arguments.

If we're willing to accept the performance cost, I agree that it makes sense to use a Python implementation of `pad` to convert arguments to an internal format and then call the C++ implementation. I still think Tensor would be the best internal format for these args.

Does that all sound good, or might there be a better approach?

**Comment by mruberry** [comment]
This sounds good to me and I think we can get started with it.

cc @heitorschueroff to see if he has some additional thoughts

**Comment by heitorschueroff** [comment]
I agree with aiming for NumPy compatibility for our Python interface only and keeping the C++ interface simpler to avoid an explosion of overloads. In this case, converting the inputs to tensors in the python wrapper is acceptable. For JIT support, you might have to tell JIT to look at the functional python interface for the types because by default it will look at native_functions.yaml (torch.tensordot does that currently). 

Maybe in the future we can accept c10::variant in native_functions.yaml for cases like this.

**Comment by kurtamohler** [comment]
Just FYI, I realized that `numpy.pad` has an undocumented acceptable format for the pad specifier type arguments (like `pad_width`, `constant_values`, etc). `((both_1), …, (both_N))` specifies the same value for both before- and after-padding, but unique for each dimension.

**Comment by mruberry** [comment]
It'dll be nice to have torch.pad which documents all these options!

> Just FYI, I realized that `numpy.pad` has an undocumented acceptable format for the pad specifier type arguments (like `pad_width`, `constant_values`, etc). `((both_1), …, (both_N))` specifies the same value for both before- and after-padding, but unique for each dimension.

It'll be nice to have torch.pad which clearly documents all these options!

**Comment by kurtamohler** [comment]
I just realized something else. `torch.nn.functional.pad` supports negative pad widths, which means that it can shorten dimensions instead of extending them, like this:
```python
>>> a = torch.arange(9).reshape((3, 3))
>>> torch.nn.functional.pad(a, (0, -1, 0, 0))
tensor([[0, 1],
        [3, 4],
        [6, 7]])
```

`numpy.pad` does not support negative pad widths. We could consider adding support for negative widths in `torch.pad`--it would be useful for the backward implementation. `torch.nn.functional.pad`'s backward function negates the padding widths and runs the pad function again to calculate the gradient.

**Comment by KunhuanLiu** [comment]
I wonder if there is any update on this issue, as I'm encountering problems pertinent to #57911.  Padding with numpy style 'wrap' could resolve my issue.
